cmd/influxd/handler.go

```diff
@@ -1,42 +1,117 @@
 package main
 
 import (
+	"log"
 	"net/http"
+	"net/url"
 	"strings"
+
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/httpd"
+	"github.com/influxdb/influxdb/messaging"
+	"github.com/influxdb/influxdb/raft"
 )
 
 // Handler represents an HTTP handler for InfluxDB node.
 // Depending on its role, it will serve many different endpoints.
 type Handler struct {
-	brokerHandler http.Handler
-	serverHandler http.Handler
+	Log    *raft.Log
+	Broker *influxdb.Broker
+	Server *influxdb.Server
+	Config *Config
 }
 
 // NewHandler returns a new instance of Handler.
-func NewHandler(bh, sh http.Handler) *Handler {
-	return &Handler{
-		brokerHandler: bh,
-		serverHandler: sh,
-	}
+func NewHandler() *Handler {
+	return &Handler{}
 }
 
 // ServeHTTP responds to HTTP request to the handler.
 func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	// Route raft and messaging paths to the broker.
-	if strings.HasPrefix(r.URL.Path, "/raft") || strings.HasPrefix(r.URL.Path, "/messaging") {
-		if h.brokerHandler == nil {
-			http.NotFound(w, r)
-			return
+	if strings.HasPrefix(r.URL.Path, "/raft") {
+		h.serveRaft(w, r)
+		return
+	}
+	if strings.HasPrefix(r.URL.Path, "/messaging") {
+		h.serveMessaging(w, r)
+		return
+	}
+
+	h.serveData(w, r)
+}
+
+// serveMessaging responds to broker requests
+func (h *Handler) serveMessaging(w http.ResponseWriter, r *http.Request) {
+	// If we're running a broker, handle the broker endpoints
+	if h.Broker != nil {
+		mh := &messaging.Handler{
+			Broker:      h.Broker.Broker,
+			RaftHandler: &raft.Handler{Log: h.Log},
 		}
+		mh.ServeHTTP(w, r)
+		return
+	}
+
+	if h.Server == nil {
```

Comment:
How could this happen? If we're not a broker, and not a server, something has gone badly wrong, right?? If we do want this check, I feel like it should be at the top of the function, i.e.

```
if h.Server == nil && h.Broker == nil {
    http.Error(w, "invalid node state", http.StatusInternalError)
    return
}
```